### PR TITLE
Add EnumClassInfo

### PR DESCRIPTION
**Pull request by UnderscoreTud**
### Description
This PR adds a subclass for ClassInfo which is targeted at enums, it automatically creates a language node for the enum, a parser, serializer, default expression and usage for the type, which should make it less tedious to make enum classinfos

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** I don't believe there are any

### CODE REVIEW

File: src/main/java/ch/njol/skript/classes/EnumClassInfo.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+				}
+
+				@Override
+				public String toVariableNameString(T o) {
+					return o.name();
+				}
+			});
+	}
+
+	/**
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/classes/EnumClassInfo.java:59]: What's this JavaDocs formatting doing? It's different from the first constructor.

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/EnumClassInfo.java (hunk 1 above)
Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/classes/EnumClassInfo.java:59]: Oh oops, mb i'll fix it rn

File: src/main/java/ch/njol/skript/classes/EnumClassInfo.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ */
+package ch.njol.skript.classes;
+
+import ch.njol.skript.expressions.base.EventValueExpression;
+import ch.njol.skript.lang.DefaultExpression;
+import ch.njol.skript.lang.ParseContext;
+import ch.njol.skript.util.EnumUtils;
+import org.eclipse.jdt.annotation.Nullable;
+
+public class EnumClassInfo<T extends Enum<T>> extends ClassInfo<T> {
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/classes/EnumClassInfo.java:27]: Add Javadocs

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/EnumClassInfo.java (hunk 2 above)
Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/classes/EnumClassInfo.java:27]: what should it be

Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/classes/EnumClassInfo.java:27]: Something describing what it's used for, what it does and how to use it

**Review by TPGamesNL — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/data/BukkitClasses.java [hunk 3]
```diff
@@ -942,45 +874,13 @@ public String toVariableNameString(InventoryHolder holder) {
...
-		Classes.registerClass(new ClassInfo<>(GameMode.class, "gamemode")
+		Classes.registerClass(new EnumClassInfo<>(GameMode.class, "gamemode", "game modes")
 				.user("game ?modes?")
 				.name("Game Mode")
 				.description("The game modes survival, creative, adventure and spectator.")
-				.usage("creative/survival/adventure/spectator")
 				.examples("player's gamemode is survival",
 						"set the player argument's game mode to creative")
-				.since("1.0")
-				.defaultExpression(new SimpleLiteral<>(GameMode.SURVIVAL, true))
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/classes/data/BukkitClasses.java:953]: The default expression here is different than the event value expression. Needs to be defined in the arguments for the EnumClassInfo constructor to be survival as default, thus making it not removed.

`new EnumClassInfo<>(GameMode.class, "gamemode", "game modes", new SimpleLiteral<>(GameMode.SURVIVAL, true))`

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/BukkitClasses.java (hunk 3 above)
Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/classes/data/BukkitClasses.java:953]: I made sure that all classes retained their default expression, this one must've slipped mb

File: src/main/java/ch/njol/skript/classes/EnumClassInfo.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	 */
+	public EnumClassInfo(Class<T> c, String codeName, String languageNode, DefaultExpression<T> defaultExpression) {
+		super(c, codeName);
+		EnumUtils<T> enumUtils = new EnumUtils<>(c, languageNode);
+		usage(enumUtils.getAllNames())
+			.serializer(new EnumSerializer<>(c))
+			.defaultExpression(defaultExpression)
+			.parser(new Parser<T>() {
+				@Override
+				public @Nullable T parse(String s, ParseContext context) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/classes/EnumClassInfo.java:49]:
```suggestion
				@Nullable
				public T parse(String s, ParseContext context) {
```

File: src/main/java/ch/njol/skript/classes/EnumClassInfo.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+				}
+			});
+	}
+
+	/**
+	 * @param c The class
+	 * @param codeName The name used in patterns
+	 * @param languageNode The language node of the type
+	 */
+	public EnumClassInfo(Class<T> c, String codeName, String languageNode) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/classes/EnumClassInfo.java:70]: I would move this constructor above the other one (funnel down into the final one)

File: src/main/java/ch/njol/skript/classes/data/BukkitClasses.java [hunk 6]
```diff
@@ -28,6 +28,9 @@
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
+import ch.njol.skript.classes.*;
+import ch.njol.skript.lang.util.SimpleLiteral;
+import ch.njol.skript.util.*;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/classes/data/BukkitClasses.java:33]: Probably preferred to avoid * imports and keep it as it was :)

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Awesome awesome stuff! Minor formatting requests :)

**Review by APickledWalrus — APPROVED** [review]

**Review by TheLimeGlass — APPROVED** [review]

**Pull request merged**